infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,195 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.sink.LogReplicationSinkEvent;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class manages the tasks submitted by the runtime FSM and the replication FSM.
+ * Both the FSMs have their own instances of this class to which tasks are submitted and executed.
```

Comment:
reason would be to keep it simple to read/maintain. I had deliberated to have 1 manager class for both FSMs, and even if we should combine both the fsm. But then concluded it would increase the code complexity and would not give any benefit in return.

Also, I tam thinking that its possible that runtime fsm event might get starved because the replicating Fsm events are more in number.

Reply:
If the thread pools are different, there will be no starvation, right?

Reply:
yes, there is no starvation for runtime events now because of replication events.

My point was incase we combined the threadpool to manage both runtime and replication events

Reply:
ohhk, yeah got your point. the code will be a bit verbose, but managing the threadpool will be simpler. I'll work on this.